[WIP] Enable EIP-155 transactions as part of FIP-0091 (#1003)
filecoin-project/FIPs: FIPS/fip-0091.md

created: 2024-04-17
---

> # FIP-0091: Add support for Homstead and EIP-155 Ethereum Transactions ("legacy" Ethereum transactions)

Reviewer: It's the second time I review this FIP and I always miss the `homstead` typos. Something wrong with my brain.

> ## Simple Summary

Reviewer: EIP155 exists and is correct (and linked in the original PR) -- it introduces chain_id as part of the tx.

EIP1155 doesn't exist, it's actually an ERC number.

> ## Simple Summary

Author: @kaitlin-beegle Hey yes as Jorge pointed out above, EIP-155 can be found at https://eips.ethereum.org/EIPS/eip-155.

}
```

> For Homestead transactions, the main distinction relevant to this FIP is that Homestead transactions do not contain a `ChainID` and so the transactions are not scoped to a specific chain. This allows replaying the same signed contract creation transaction across multiple chains for deploying a contract and also enables the contract to have the same address across chains.

Reviewer 2: ```suggestion
For Homestead transactions, the main distinction relevant to this FIP is that Homestead transactions do not contain a `ChainID`, so the transactions are not scoped to a specific chain. This allows replaying the same signed contract creation transaction across multiple chains for deploying a contract and also enables the contract to have the same address across chains.
```

> EIP-155 transactions do not have an explicit `ChainID` parameter but they are scoped to a specific chain and enable chain replay protection by encoding the `ChainID` in the `V` parameter of the signature and ensuring that the transaction digest signed by the user contains the `ChainID`. EIP-155 transactions allow users of wallets like Coinbase Wallet to send EVM transactions to Filecoin as these wallets use EIP-155 EVM transactions.

Reviewer 2: ```suggestion
EIP-155 transactions do not have an explicit `ChainID` parameter but they are scoped to a specific chain and enable chain replay protection by encoding the `ChainID` in the `V` parameter of the signature, ensuring that the transaction digest signed by the user contains the `ChainID`. EIP-155 transactions allow users of wallets like Coinbase Wallet to send EVM transactions to Filecoin as these wallets use EIP-155 EVM transactions.
```

Support for legacy transactions on Filecoin will be enabled by prepending an extra marker byte to the signature of legacy transactions. This will allow Filecoin to detect that a transaction is a legacy transaction and process and execute it accordingly. For Homestead transactions, the marker byte will be `0x01` and for EIP-155 transactions, the marker byte will be `0x02`.

The `GasPrice` parameter from the legacy transaction will be utilized to determine both the `GasFeeCap` and `GasPremium` parameters in the Filecoin message, as `GasPrice` encompasses the full cost per unit of gas that the sender is willing to pay. This logic will be the same for both Homestead transactions and EIP-155 transactions.

> To ensure compatibility with Filecoin's signature verification, which only recognizes `V` values of 0 or 1 (as is also the case with the current signature verification implementation in `go-ethereum`), adjustments will be made to the `V` parameter in legacy transactions before verifying their signatures. Specifically, for Homestead transactions, 27 will be subtracted from the `V` parameter. For EIP-155 transactions, the adjustment will involve subtracting (2*ChainID + 35) from the `V` parameter.
> 
> This adjustment is necessary because, unlike EIP-1559 transactions supported by Filecoin that strictly use `V` values of 0 or 1, the `V` parameter in Homestead transactions can be either 27 or 28, and for EIP-155 transactions, it is calculated as (0 or 1 + (2 * ChainID + 35)) where the selection of 0 or 1 is the same as in EIP-1599 transactions.

Reviewer 3: ```suggestion
To ensure compatibility with Filecoin's signature verification, which only recognizes `V` values of 0 or 1 (as is also the case with the current signature verification implementation in `go-ethereum`), adjustments will be made to the `V` parameter in legacy transactions before verifying their signatures. 

For Homestead transactions, the 'V' parameter will be calculated as _V_-27. This is because Homestead transactions are coded as either 27 or 28. 

For EIP-155 transactions, the adjustment will involve subtracting (2*ChainID + 35) from the `V` parameter.  This is because it is calculated as (0 or 1 + (2 * ChainID + 35)) where the selection of 0 or 1 is the same as in EIP-1599 transactions.
```

Author: I think the original more descriptive language will help flush out better the change that needs to be made here.

To ensure chain replay protection, a Filecoin node must verify that the `ChainID` signed by the user in the transaction digest matches the `ChainID` of the network to which the node is connected.

To differentiate a Filecoin message representing an EIP-155 transaction from a Filecoin message representing an EIP-1559 or a Homestead transaction, and to accurately reconstruct the original EIP-155 transaction from the Filecoin message, this FIP proposes prepending a distinct marker byte `0x02` to the signature of an EIP-155 transaction. This adjustment takes place after the EIP-155 transaction is submitted to a Filecoin node but before it is broadcast across the network.

As mentioned earlier, it is easy to distinguish whether a newly submitted legacy ETH transaction is an EIP-155 transaction or a Homestead transaction based on the `V` value of the signature. The marker byte can be prepended to the signature accordingly before broadcasting it across the network.

> The signature length for a Filecoin message representing an EIP-155 transaction can be calculated as (1 byte for the marker byte `0x02` + 32 bytes for the `R` value of the ETH signature + 32 bytes for the `S` value of the ETH signature + size of `V` value), where `V` is (0 or 1 + (2 * ChainID + 35)). Thus, the signature length depends on the size of the `ChainID`. However, since the `ChainID` is known at runtime, the signature length of a Filecoin message representing an EIP-155 transaction is also known at runtime.

Reviewer 4: Can you say in here what it is, for mainnet at least? `314` I guess gives us 2 bytes so it should be `67` long?  There's no situation in which it could be 1 byte long which, I think, would give us the same length as a homestead transaction? It might be worth adding some wording in here about the likely difference between the length of the two given that the largest chain id we have is mainnet which is >1 byte so length comparisons as an initial branching factor _should_ be safe.

Author: @rvagg Good point. I have pushed this change.

Reviewer 4: nice, thanks!
